gocourse04/SmartZoo/main.go

```diff
@@ -0,0 +1,255 @@
+для перебування тварини і технічне приміщення, де можуть зберігатися приладдя для роботи з тваринами. На кожній
+території є певна кількість тварин, а технічне приміщення «здатне» виконувати деякі функції, такі як прибирання і
+годування тварин. Реалізувати функції пошуку тварини за імʼям або ID, переміщення тварини із загону в загін,
+годування тварини.
+*/
+package main
+
+import (
+	"errors"
+	"fmt"
+)
+
+const (
+	areaClean   = "tech room"
+	sectorClean = 6
+)
+
+type Areas map[string]Area
+
+// Територія по  типу (копитні, пернаті, примати)
+type Area struct {
+	Name    string
+	Sectors map[string]Sector
+}
+
+// Загон для виду тварин. У кожному загоні є технічне приміщення (наприклад годування)
+type Sector struct {
+	ID      int
+	Animals []Animal
+}
+
+type Animal struct {
+	ID   int
+	Name string
+}
+
+type Zoo struct {
+	Areas Areas
+}
+
+func NewZoo(areasData map[string]Area) *Zoo {
+	return &Zoo{Areas: areasData}
+}
+
+var AreasData = map[string]Area{
+	"Hoofed": {
+		Name: "hoofed", Sectors: map[string]Sector{
+			"Horses": {
+				ID: 1,
+				Animals: []Animal{
+					{ID: 1, Name: "Horse1"},
+					{ID: 2, Name: "Horse2"},
+					{ID: 3, Name: "Horse3"},
+				},
+			},
+			"Cows": {
+				ID: 2,
+				Animals: []Animal{
+					{ID: 4, Name: "Cow1"},
+					{ID: 5, Name: "Cow2"},
+					{ID: 6, Name: "Cow3"},
+				},
+			},
+		},
+	},
+	"Birds": {
+		Name: "birds", Sectors: map[string]Sector{
+			"Doves": {
+				ID: 3,
+				Animals: []Animal{
+					{ID: 7, Name: "Dove1"},
+					{ID: 8, Name: "Dove2"},
+					{ID: 9, Name: "Dove3"},
+				},
+			},
+			"Hawks": {
+				ID: 4,
+				Animals: []Animal{
+					{ID: 10, Name: "Hawk1"},
+					{ID: 11, Name: "Hawk2"},
+					{ID: 12, Name: "Hawk3"},
+				},
+			},
+		},
+	},
+	"Techroom": {
+		Name: "tech room", Sectors: map[string]Sector{
+			"Dining room": {
+				ID: 5,
+				Animals: []Animal{
+					{ID: 13, Name: "Hawk4"},
+				},
+			},
+			"Bath room": {
+				ID: 6,
+				Animals: []Animal{
+					{ID: 14, Name: "Cow4"},
+				},
+			},
+		},
+	},
+}
+
+// Lookup returns the found animal and in which sector and area it is.
+func (z *Zoo) Lookup(name string) (*Area, *Sector, *Animal) {
+	for _, area := range z.Areas {
+		for _, sector := range area.Sectors {
+			for _, animal := range sector.Animals {
+				if animal.Name == name {
+					return &area, &sector, &animal
+				}
+			}
+		}
+	}
+	return nil, nil, nil
+}
+
+// move the animal to another sector
+func (z *Zoo) Migration(fromAreaName, toArea string, fromSectorID, toSector int, animal Animal) error {
+	for _, area := range z.Areas {
+		if area.Name == toArea {
+			for sectorName, sector := range area.Sectors {
+				if sector.ID == toSector {
+					sector.Animals = append(sector.Animals, animal)
+					area.Sectors[sectorName] = sector // rewrite sector
+					z.deleteOldRecord(fromAreaName, fromSectorID, animal)
+					return nil
+				}
+			}
+		}
+	}
+	return fmt.Errorf("sector not found")
```

Comment:
```suggestion
	return errors.New("sector not found")
```

`fmt.Errorf` - це якщо треба щось додати до меседжа помилки